ai_ta_backend/vector_database.py

```diff
@@ -1081,18 +1083,96 @@ def getTopContexts(self, search_query: str, course_name: str, token_limit: int =
+    2. Retrieve relevant docs for each query.
+    3. Rank the docs based on the relevance score.
+    4. Pad the top 5 docs with context from the original document.
+    """
+    try:
+      top_n = 80 # HARD CODE TO ENSURE WE HIT THE MAX TOKENS
+      start_time_overall = time.monotonic()
+
+      # Vector search with ONLY original query
+      # found_docs: list[Document] = self.vector_search(search_query=search_query, course_name=course_name)
+      mq_start_time = time.monotonic()
+      # Multi query retriever
+      generate_queries = (
+                MULTI_QUERY_PROMPT
+                | self.llm
+                | StrOutputParser()
+                | (lambda x: x.split("\n"))
+                | (lambda x: list(filter(None, x)))  # filter out non-empty strings
+            )
       
-      # call context padding function here
-      final_docs = self.context_padding(found_docs, search_query, course_name)
+      generated_queries = generate_queries.invoke({"original_query": search_query})
+      print("generated_queries", generated_queries)
+
+      batch_found_docs: list[list[Document]] = self.batch_vector_search(search_queries=generated_queries, course_name=course_name)
+
+      found_docs = self.reciprocal_rank_fusion(batch_found_docs)
+      found_docs = [doc for doc, score in found_docs]
+      print(f"Number of docs found with multiple queries: {len(found_docs)}")
+      if len(found_docs) == 0:
+          return []
+
+      print(f"⏰ Multi-query processing runtime: {(time.monotonic() - mq_start_time):.2f} seconds")
+
+      # Context filtering
+      #filtered_docs = list(run(contexts=found_docs, user_query=search_query, max_time_before_return=45, max_concurrency=100))
+      #print(f"Number of docs after context filtering: {len(filtered_docs)}")
+
+      # 'context padding' // 'parent document retriever' 
+      final_docs = context_processing(found_docs, search_query, course_name)
+      print(f"Number of final docs after context padding: {len(final_docs)}")
 
       pre_prompt = "Please answer the following question. Use the context below, called your documents, only if it's helpful and don't use parts that are very irrelevant. It's good to quote from your documents directly, when you do always use Markdown footnotes for citations. Use react-markdown superscript to number the sources at the end of sentences (1, 2, 3...) and use react-markdown Footnotes to list the full document names for each number. Use ReactMarkdown aka 'react-markdown' formatting for super script citations, use semi-formal style. Feel free to say you don't know. \nHere's a few passages of the high quality documents:\n"
       # count tokens at start and end, then also count each context.
       token_counter, _ = count_tokens_and_cost(pre_prompt + '\n\nNow please respond to my query: ' + search_query) # type: ignore
 
+      filtered_docs = ray_run(contexts=final_docs, user_query=search_query, max_time_before_return=45, max_concurrency=100)

```

Comment:
I think Rohan and I hypothesized that we should filter relevant docs _after_ the multiquery and _before_ the reciprocal rerank function. Do you think that's a better or worse place for filtering than here, at the end? Both have pros and cons. Thanks.

Reply:
My logic behind placing it at the end was to create a fail-safe, if it doesn't pass any contexts as "useful" to answer the query, we can directly use the contexts obtained before the filtering step. The number of contexts it filters varies depending on the question - I have seen it pass anything from 0 to 40 contexts. 
Your hypothesis is also correct, if we filter it at the start itself, we have less work to do in ranking and padding. 

What do we do if it passes 0 contexts after filtering? Currently, I am using the ones before the filtering step for this condition.

Reply:
Update: I have added a condition for this. If it passes less than 5 contexts, use the pre-filtered version, else use the filtered contexts.